File: test/org/cerion/symcalc/expression/function/list/JoinTest.java
package org.cerion.symcalc.expression.function.list;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.junit.Test;

import static org.junit.Assert.*;

public class JoinTest {

    private static final ListExpr L1 = new ListExpr(IntegerNum.ZERO, IntegerNum.ONE);
    private static final ListExpr L2 = new ListExpr(IntegerNum.TWO);

    @Test
    public void validate() {
        Expr e1 = new ListExpr();
        Expr e2 = IntegerNum.TWO;
        Expr e3 = new ListExpr(IntegerNum.ZERO, IntegerNum.ONE);

        assertTrue( new Join(e1, e2).eval().isError());
        assertTrue( new Join(e2).eval().isError());
        assertTrue( new Join(e2, e3).eval().isError());
    }

    @Test
    public void emptyList() {
        ListExpr e = new ListExpr();
        assertEquals(e, new Join(e).eval());
        assertEquals(e, new Join(e,e).eval());
    }

    @Test
    public void basicTest() {
        Expr e = new Join(L1, L2).eval();

        ListExpr expected = new ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO);
        assertEquals(expected, e);
    }
}

File: test/expression/function/list/JoinTest.kt
package org.cerion.symcalc.expression.function.list

import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.number.Integer
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class JoinTest {

    @Test
    fun validate() {
        val e1 = ListExpr()
        val e2 = Integer.TWO
        val e3 = ListExpr(Integer.ZERO, Integer.ONE)

        assertTrue(Join(e1, e2).eval().isError)
        assertTrue(Join(e2).eval().isError)
        assertTrue(Join(e2, e3).eval().isError)
    }

    @Test
    fun emptyList() {
        val e = ListExpr()
        assertEquals(e, Join(e).eval())
        assertEquals(e, Join(e, e).eval())
    }

    @Test
    fun basicTest() {
        val e = Join(L1, L2).eval()

        val expected = ListExpr(Integer.ZERO, Integer.ONE, Integer.TWO)
        assertEquals(expected, e)
    }

    companion object {
        private val L1 = ListExpr(Integer.ZERO, Integer.ONE)
        private val L2 = ListExpr(Integer.TWO)
    }
}